{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import urllib.request\n",
    "\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins.dada2.methods import denoise_single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_file(url, local_filepath):\n",
    "    \"\"\"Download a file from a remote url and save to a local filepath\n",
    "    \n",
    "    url - the web address of the file you want to download as a string\n",
    "    local_filepath - the local filepath to which the file will be saved\n",
    "    \"\"\"\n",
    "    print(f\"Downloading file: {url}\")\n",
    "    with urllib.request.urlopen(url) as response, open(local_filepath, 'wb') as out_file:\n",
    "        shutil.copyfileobj(response, out_file)\n",
    "    \n",
    "    print(f\"Saved to local filepath: {local_filepath}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_directory(path):\n",
    "    \"\"\"Make a directory, but proceed without errors if it fails\n",
    "    path -- the path to the directory (e.g. \"../output/taxonomy_references\")\n",
    "    \"\"\"\n",
    "    try:\n",
    "        os.mkdir(path)\n",
    "    except OSError:\n",
    "        print (f\"Creation of directory {path} failed\")\n",
    "    else:\n",
    "        print (f\"Created the directory {path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_dir=os.path.abspath('../')\n",
    "mock_dir=working_dir+'/output/mock_communities'\n",
    "metadata_url_base='https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mocks=[12, 13, 14, 15, 16, 18, 19, 20, 21, 22]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating mock directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_12\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_12\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-12/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-12/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-12/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_12/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-jjsuduaa/ed8f0304-ab58-468e-a001-c50bf7c67749/data /tmp/tmpju4nca60/output.tsv.biom /tmp/tmpju4nca60/track.tsv /tmp/tmpju4nca60 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 12 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_13\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_13\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-13/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-13/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-13/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_13/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-ssxhoas5/cf9e7237-eb94-4b4d-9b94-d0b02870bfa3/data /tmp/tmp_65_a9ur/output.tsv.biom /tmp/tmp_65_a9ur/track.tsv /tmp/tmp_65_a9ur 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 13 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_14\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_14\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-14/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-14/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-14/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_14/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-h_p2zb5q/c919744b-1f0b-40c4-98be-a9fbfa5b22e3/data /tmp/tmpu70f0k3r/output.tsv.biom /tmp/tmpu70f0k3r/track.tsv /tmp/tmpu70f0k3r 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 14 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_15\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_15\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-15/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-15/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-15/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_15/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-k9ae6dt7/b41e04ef-ca58-498f-be21-e900da79e8ff/data /tmp/tmpco7f0enx/output.tsv.biom /tmp/tmpco7f0enx/track.tsv /tmp/tmpco7f0enx 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 15 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_16\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_16\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-16/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-16/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-16/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_16/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-39c7ucsq/1ca241b4-0feb-4bd2-bb1c-dcf74ce188cd/data /tmp/tmplrapencj/output.tsv.biom /tmp/tmplrapencj/track.tsv /tmp/tmplrapencj 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mock 16 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_18\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_18\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-18/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-18/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-18/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_18/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-4xhxo3ab/c92daa12-99bc-400e-9078-612926b1b18f/data /tmp/tmphnn4u9h3/output.tsv.biom /tmp/tmphnn4u9h3/track.tsv /tmp/tmphnn4u9h3 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 18 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_19\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_19\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-19/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-19/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-19/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_19/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-tdciaz3b/b810f35a-435d-4e5c-85b0-04730e39d8da/data /tmp/tmph47qo43p/output.tsv.biom /tmp/tmph47qo43p/track.tsv /tmp/tmph47qo43p 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 19 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_20\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_20\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-20/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-20/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-20/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_20/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-ip8jlqly/814cc117-faf3-492e-b1ea-84aa8f01f476/data /tmp/tmpnmam40ll/output.tsv.biom /tmp/tmpnmam40ll/track.tsv /tmp/tmpnmam40ll 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 20 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_21\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_21\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-21/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-21/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-21/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_21/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-qgb3zgqe/8354eefa-48f2-479c-b480-50725e75da66/data /tmp/tmp683cnmcw/output.tsv.biom /tmp/tmp683cnmcw/track.tsv /tmp/tmp683cnmcw 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 21 imported\n",
      "\n",
      "Creating mock folder /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_22\n",
      "Created the directory /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_22\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-22/dataset-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/metadata.tsv\n",
      "Downloading file: https://github.com/caporaso-lab/mockrobiota/raw/master/data/mock-22/sample-metadata.tsv\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/sample_metadata.tsv\n",
      "Downloading file: https://s3-us-west-2.amazonaws.com/mockrobiota/latest/mock-22/mock-forward-read.fastq.gz\n",
      "Saved to local filepath: /mnt/c/Users/Dylan/Documents/zaneveld/2_14_gcmp/GCMP_Global_Disease/analysis/organelle_removal/output/mock_communities/mock_22/1_1_L001_R1_001.fastq.gz\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-fpzvdppb/66567855-e2ef-41eb-a6eb-6ad3c9f4ea60/data /tmp/tmpsqjihi0_/output.tsv.biom /tmp/tmpsqjihi0_/track.tsv /tmp/tmpsqjihi0_ 0 0 2.0 2 Inf independent consensus 1.0 0 1000000 NULL 16\n",
      "\n",
      "mock 22 imported\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#download the forward reads, import into qiime2, and create\n",
    "#sequence and feature table artifacts\n",
    "#this loop could take a while, depending on the power of your\n",
    "#cpu and your download speed\n",
    "if not os.path.exists(mock_dir):\n",
    "    print(f'Creating mock directory {mock_dir}')\n",
    "    make_directory(mock_dir)\n",
    "for mock in mocks:\n",
    "    mock_folder=mock_dir+'/mock_'+str(mock)\n",
    "    if not os.path.exists(mock_folder):\n",
    "        print(f'Creating mock folder {mock_folder}')\n",
    "        make_directory(mock_folder)\n",
    "    metadata_url=metadata_url_base+str(mock)+'/dataset-metadata.tsv'\n",
    "    metadata_filepath=mock_dir+'/metadata.tsv'\n",
    "    sample_metadata_url=metadata_url_base+str(mock)+'/sample-metadata.tsv'\n",
    "    sample_metadata_filepath=mock_dir+'/sample_metadata.tsv'\n",
    "    download_file(metadata_url, metadata_filepath)\n",
    "    download_file(sample_metadata_url, sample_metadata_filepath)\n",
    "    metadata=pd.read_csv(metadata_filepath, sep='\\t')\n",
    "    sample_metadata=pd.read_csv(sample_metadata_filepath, sep='\\t')\n",
    "    sample_name=sample_metadata['SampleID'][0]\n",
    "    forward_url=metadata['value'].values[2]\n",
    "    forward_filepath=mock_folder+'/1_1_L001_R1_001.fastq.gz'\n",
    "    #reverse_url=metadata['value'].values[3]\n",
    "    #reverse_filepath=mock_folder+'/'+str(mock)+'_reverse.fastq.gz'\n",
    "    #index_url=metadata['value'].values[4]\n",
    "    #index_filepath=mock_folder+'/'+str(mock)+'_index.'\n",
    "    download_file(forward_url, forward_filepath)\n",
    "    #download_file(reverse_url, reverse_filepath)\n",
    "    #importing a fastq file into qiime2 requires a manifest file and a metadata.yml file\n",
    "    if not os.path.exists(mock_folder+'/MANIFEST'):\n",
    "        with open(mock_folder+'/MANIFEST', 'w') as file:\n",
    "            file.write('sample-id,filename,direction\\n'+'sample1,'+forward_filepath+',forward')\n",
    "    if not os.path.exists(mock_folder+'/metadata.yml'):\n",
    "        with open(mock_folder+'/metadata.yml', 'w') as file:\n",
    "            pass\n",
    "    #now we need to import them into qiime\n",
    "    seqs = Artifact.import_data('SampleData[SequencesWithQuality]', '../output/mock_communities/mock_'+str(mock))\n",
    "    seqs.save(mock_folder+'/mock_'+str(mock)+'_sequences.qza')\n",
    "    table, sequences, stats=denoise_single(seqs, 0, n_threads=0)\n",
    "    table.save(mock_folder+'/mock_'+str(mock)+'_ft.qza')\n",
    "    sequences.save(mock_folder+'/mock_'+str(mock)+'_rep_seqs.qza')\n",
    "    print('mock', mock, 'imported\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
